### PR TITLE
BP-2381 Beez mods

### PR DESCRIPTION
**Pull request by corkupine**

### CODE REVIEW

File: MsmqReturnToSourceQueue/ErrorManager.cs [hunk 1]
```diff
@@ -125,6 +134,21 @@ public void ReturnMessageToSourceQueue(string messageId)
         }
     }
 
+    private void AddOrIncrementReturnToSourceQueueCount(Message message, Dictionary<string, string> headers)
+    {
+        var returnToSourceQueueCountExists = headers.ContainsKey(ReturnToSourceQueueCount);
+        if (returnToSourceQueueCountExists)
+        {
+            var returnToSourceQueueCount = int.Parse(headers[ReturnToSourceQueueCount]) + 1;
```
Comment (derfsplat) [inline MsmqReturnToSourceQueue/ErrorManager.cs]: Worth a `TryParse` here?  Consider `Convert.ToIn32` as that doesn't throw on null
https://stackoverflow.com/questions/199470/whats-the-main-difference-between-int-parse-and-convert-toint32

File: MsmqReturnToSourceQueue/MsmqAddress.cs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 using System.Net;
 
 
-internal struct MsmqAddress
+public struct MsmqAddress
```
Comment (derfsplat) [inline MsmqReturnToSourceQueue/MsmqAddress.cs:5]: why did this need to change?  The entire project should be able to reference something that's `internal`

File: MsmqReturnToSourceQueue/MsmqUtilities.cs [hunk 3]
```diff
@@ -7,15 +7,23 @@
 using System.Net.NetworkInformation;
 using System.Text;
 using System.Xml;
+using System.Xml.Linq;
 
-class MsmqUtilities
+public class MsmqUtilities
```
Comment (derfsplat) [inline MsmqReturnToSourceQueue/MsmqUtilities.cs:12]: super minor but in the context of `MsMqAddress` i'd say this should be `internal`

File: MsmqReturnToSourceQueue/ErrorManager.cs [hunk 4]
```diff
@@ -89,7 +98,7 @@ public void ReturnMessageToSourceQueue(string messageId)
 
                         if (string.IsNullOrEmpty(originalId) || messageId != originalId)
                         {
-                            if (messageCount%ProgressInterval == 0)
+                            if (messageCount % ProgressInterval == 0)
```
Comment (elovelan) [inline MsmqReturnToSourceQueue/ErrorManager.cs]: please keep refactors in separate commits. This helps if things need to be reverted

File: MsmqReturnToSourceQueue/ErrorManager.cs [hunk 5]
```diff
@@ -140,13 +164,4 @@ string GetOriginalId(Dictionary<string, string> headers)
 
         return null;
     }
-
-
-    const string NoMessageFoundErrorFormat = "INFO: No message found with ID '{0}'. Checking headers of all messages.";
```
Comment (derfsplat) [inline MsmqReturnToSourceQueue/ErrorManager.cs]: if we ever intend to submit our changes back to the original repo, moving declarations to suit a different style would be considered bad form:
http://tirania.org/blog/archive/2010/Dec-31.html
> When you contribute fixes or new features to an open source project you should use the existing coding style, the existing coding patterns and stick by the active maintainer's choice for his code organization.

File: MsmqReturnToSourceQueue/MsmqUtilities.cs [hunk 6]
```diff
@@ -179,21 +181,26 @@ static string GetCorrelationId(Message message, Dictionary<string, string> heade
         return result;
     }
 
-    public static string GetFullPathWithoutPrefix(MsmqAddress address)
+    public static void SaveMessageHeaders(Dictionary<string, string> headers, Message m)
+    {
+        var root = new XElement("ArrayOfHeaderInfo");
+        foreach (var header in headers)
+        {
+            var headerInfo = new XElement("HeaderInfo");
```
Comment (derfsplat) [inline MsmqReturnToSourceQueue/MsmqUtilities.cs]: other magic strings in this class are constants

**Review by derfsplat — CHANGES_REQUESTED** [review]
reviewed :orange_book:

File: MsmqReturnToSourceQueue/MsmqUtilities.cs [hunk 7]
```diff
@@ -179,21 +181,26 @@ static string GetCorrelationId(Message message, Dictionary<string, string> heade
         return result;
     }
 
-    public static string GetFullPathWithoutPrefix(MsmqAddress address)
+    public static void SaveMessageHeaders(Dictionary<string, string> headers, Message m)
```
Comment (elovelan) [inline MsmqReturnToSourceQueue/MsmqUtilities.cs:183]: For consistency, I would use the same mechanisms as in the [original source for this file](https://github.com/Particular/NServiceBus/blob/5.2.25/src/NServiceBus.Core/Transports/Msmq/MsmqUtilities.cs#L246-L254)

File: MsmqReturnToSourceQueue.sln.DotSettings [hunk 8]
```diff
@@ -557,11 +558,19 @@ II.2.12 &lt;HandlesEvent /&gt;&#xD;
...
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateThisQualifierSettings/@EntryIndexedValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsCodeFormatterSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsParsFormattingSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002EJavaScript_002ECodeStyle_002ESettingsUpgrade_002EJsWrapperSettingsUpgrader/@EntryIndexedValue">True</s:Boolean>
```
Comment (elovelan) [inline MsmqReturnToSourceQueue.sln.DotSettings:573]: what does adding all of these do?

File: MsmqReturnToSourceQueue/Program.cs [hunk 9]
```diff
@@ -42,7 +46,7 @@ static void Main(string[] args)
         {
             Console.WriteLine("Input queue [{0}] resides on a remote machine: [{1}].", errorQueueAddress.Queue, errorQueueAddress.Machine);
             Console.WriteLine("Due to networking load, it is advised to refrain from using ReturnToSourceQueue on a remote error queue, unless the error queue resides on a clustered machine.");
-            if (!script)
+            if (!script && !skipLocalCheck)
```
Comment (elovelan) [inline MsmqReturnToSourceQueue/Program.cs]: why is this needed?

File: MsmqReturnToSourceQueue.sln.DotSettings (hunk 8 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue.sln.DotSettings:573]: Beats me - VS did it.

File: MsmqReturnToSourceQueue/MsmqAddress.cs (hunk 2 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/MsmqAddress.cs:5]: Tests

File: MsmqReturnToSourceQueue/MsmqUtilities.cs (hunk 3 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/MsmqUtilities.cs:12]: Tests

File: MsmqReturnToSourceQueue/ErrorManager.cs (hunk 1 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/ErrorManager.cs]: It is pretty much impossible for it to not be an integer based upon logic. `<sinceroid> Is it still best practice to guard against null in this case? </sinceroid>`

File: MsmqReturnToSourceQueue/Program.cs (hunk 9 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/Program.cs]: The ability to skip the prompt and ReadKey() allows us to preclude user input.

File: MsmqReturnToSourceQueue/ErrorManager.cs (hunk 4 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/ErrorManager.cs]: I didn't think it was a big deal for what is probably a one-off modification.

Comment (corkupine) [reply MsmqReturnToSourceQueue/ErrorManager.cs]: Dammit I fixed it and it keeps sneaking in every time I save. IMO this should not hold up a merge.

**Comment by corkupine** [comment]
@derfsplat @elovelan Addressed concerns

File: MsmqReturnToSourceQueue/MsmqUtilities.cs (hunk 7 above)
Comment (corkupine) [reply MsmqReturnToSourceQueue/MsmqUtilities.cs:183]: done

File: MsmqReturnToSourceQueue/Program.cs (hunk 9 above)
Comment (elovelan) [reply MsmqReturnToSourceQueue/Program.cs]: isn't that what the `script` var is for?

File: MsmqReturnToSourceQueue.sln.DotSettings (hunk 8 above)
Comment (elovelan) [reply MsmqReturnToSourceQueue.sln.DotSettings:573]: https://www.jetbrains.com/help/resharper/Sharing_Configuration_Options.html#upgrading

File: MsmqReturnToSourceQueue/Program.cs (hunk 9 above)
Comment (elovelan) [reply MsmqReturnToSourceQueue/Program.cs]: oh, nevermind, I see now that you added this so that we could force it to prompt, even if `inputQueue` is provided. You may want to invert this since it changes the default behavior (i.e. it would previously skip the prompt if the `inputQueue` arg was provided, now it requires `inputQueue` and `skipLocalCheck` on the command-line to skip the prompt, maybe rename it to `promptWhenRemoteQueue` or similar?)

Comment (corkupine) [reply MsmqReturnToSourceQueue/Program.cs]: Oh, you're right. I missed that script was true by default.

**Comment by corkupine** [comment]
Yo @derfsplat ...take another look!

File: MsmqReturnToSourceQueue/MsmqAddress.cs (hunk 2 above)
Comment (derfsplat) [reply MsmqReturnToSourceQueue/MsmqAddress.cs:5]: in the interest of contributing this back, why don't we do this instead:
https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.internalsvisibletoattribute(v=vs.110).aspx

That way we don't have to change access modifiers.  Normally I'd say nbd, but i think we'll get gruff for this too if we want to contribute back to nsb.

Comment (corkupine) [reply MsmqReturnToSourceQueue/MsmqAddress.cs:5]: I think it doesn't matter - this is deprecated and not really in use.

**Review by derfsplat — APPROVED** [review]
